tests/data_sources/test_uspto.py

```diff
@@ -0,0 +1,494 @@
+        )
+
+    def test_child(self):
+        self.assertEqual(
+            self.uspto.tables_transitive_closure(
+                ["icpr_classifications"], "us_patents"
+            ),
+            set(["us_patents", "icpr_classifications"]),
+        )
+
+
+class TestUsptoQuery(unittest.TestCase):
+    """Verify column specification and population of root table"""
+
+    @classmethod
+    def setUpClass(cls):
+        FileCache.parse_counter = 0
+        ZipCache.file_reads = 0
+        populate_attached()
+
+        # debug.set_flags(["sql"])
+        cls.uspto = uspto.Uspto(
+            td("data/April 2023 Patent Grant Bibliographic Data"),
+            attach_databases=[f"attached_uspto:{ATTACHED_DATABASE_PATH}"],
+        )
+
+    @classmethod
+    def tearDownClass(cls):
+        del cls.uspto
+        os.unlink(ATTACHED_DATABASE_PATH)
+
+    def test_patents(self):
+        for partition in True, False:
+            self.assertEqual(
+                record_count(
+                    self.uspto.query("SELECT * FROM us_patents", partition)
+                ),
+                14,
+            )
+        self.assertEqual(FileCache.parse_counter, 28)
+        FileCache.parse_counter = 0
+
+    def test_cache(self):
+        self.assertEqual(
+            record_count(self.uspto.query("SELECT * FROM us_patents LIMIT 1")),
+            1,
+        )
+        self.assertEqual(FileCache.parse_counter, 1)
+        self.assertEqual(ZipCache.file_reads, 1)
+        FileCache.parse_counter = 0
+        ZipCache.file_reads = 0
+
+    def test_patents_attached(self):
+        for partition in True, False:
+            self.assertEqual(
+                record_count(
+                    self.uspto.query(
+                        """SELECT * FROM us_patents WHERE EXISTS (
+                SELECT 1 FROM attached_uspto.s_us_patents WHERE us_patents.type = s_us_patents.type)
+            """,
+                        partition,
+                    )
+                ),
+                4,
+            )
+        self.assertEqual(FileCache.parse_counter, 28)
+        FileCache.parse_counter = 0
+
+    def test_icpr_classifications(self):
+        for partition in True, False:
+            self.assertEqual(
+                record_count(
+                    self.uspto.query(
+                        "SELECT * FROM icpr_classifications", partition
+                    )
+                ),
+                30,
+            )
+        self.assertEqual(FileCache.parse_counter, 27)
+        FileCache.parse_counter = 0
+
+    def test_patents_claims_condition(self):
+        for partition in True, False:
+            self.assertEqual(
+                record_count(
+                    self.uspto.query(
+                        "select us_patents.claims_number from us_patents left join icpr_classifications on us_patents.container_id = icpr_classifications.patent_id where us_patents.type = 'utility' AND icpr_classifications.symbol_position = 'F'",
```

Comment:
Please use capital case for SQL keywords (throught) and format your queries to avoid them exceeding the line length (use multiline strings for this).